File: ruby/BST.rb
class BST
  attr_accessor :value,:left,:right
   def initialize(value)
    @value = value
    @left = nil
    @right =nil
  end
# O(log(n)):time | O(1) space
  def insert(value)
    current = self
    while true
       if current.value > value
        if current.left.nil?
         current.left = BST.new(value)
         break
        else
          current = current.left
        end
      else
        if current.right.nil?
          current.right= BST.new(value)
          break
        else
          current = current.right
        end
       end
    end
  end
# O(log(n)):time | O(1) space
  def contains(value)
    current = self
    while !current.nil?
      if current.value > value
        current = current.left
      elsif current.value < value
        current = current.right
      else
        return true
      end
    end
    return false
  end
  
  def getMinValue
    current = self
    until current.left.nil? 
      current = current.left
    end 
    current.value   
  end
  
  def remove(value,parent = nil)
    current = self
    if self.contains(value)
      
      while !current.nil? 

        if value < current.value
          parent = current
          current = current.left
        elsif value > current.value
          parent = current
          current = current.right
        else
          if !current.left.nil? && !current.right.nil?
            current.value = current.right.getMinValue()
            current.right.remove(current.value,current)
          elsif parent.nil?
            if !current.left.nil?
              current.value = current.left.value
              current.right = current.left.right
              current.left =current.left.right
            elsif !current.right.nil?
                 current.value = current.right.value
              current.left = current.right.right
              current.right =current.right.right
            end
          end
        elsif parrent.left == current
          if !current.left.nil
             parrent.left = current.left
          else
            parrent.left =current.right
          end
      
        elsif parrent.rigth == current
          if !current.left.nil
            parrent.left = current.left
         else
           parrent.left =current.right
         end
     
      end

    else
      return "This node does not contain this value"
    end
  end
  
end

bst = BST.new(3)
 bst.insert(4)
 bst.insert(5)
 bst.insert(7)
 bst.insert(2)
 p bst.remove(2)
# p bst

File: Js/BST.js
class BST {
  constructor(value) {
    this.value = value
    this.left = null
    this.right = null
  }

  insert(value) {
    let current = this
    while (true) {
      if (value < current.value) {
        if (current.left === null) {
          current.left = new BST(value)
          break
        } else {
          current = current.left
        }
      } else {
        if (current.right === null) {
          current.right = new BST(value)
          break
        } else {
          current = current.right
        }
      }
    }
  }
  contains(value) {
    let current = this
    while (current !== null) {
      if (value < current.value) {
        current = current.left
      } else if (value > current.value) {
        current = current.right
      } else {
        return true
      }
    }
    return false
  }

  getMin() {
    let current = this
    while (current.left !== null) {
      current = current.left
    }
    return current.value
  }

}
const bst = new BST(3)

bst.insert(2)
bst.insert(4)
bst.insert(6)
let res = bst.contains(20)
let min = bst.getMin()
console.log(min);
console.log(bst);